### PR TITLE
Define Swift `Credentials` as an enum

### PR DESCRIPTION
**Pull request by ericjordanmossman**
The purpose of this pull request is to define `Credentials` as an enumerable instead of a typealias of `RLMCredentials`.

This branch adds an ObjectCSupport.convert for `Credentials` to `RLMCredentials`.

### CODE REVIEW

File: RealmSwift/Tests/ObjectiveCSupportTests.swift [hunk 1]
```diff
@@ -95,4 +95,26 @@ class ObjectiveCSupportTests: TestCase {
                        ObjectiveCSupport.convert(object: realm.configuration).deleteRealmIfMigrationNeeded,
                        "Configuration.deleteRealmIfMigrationNeeded must be equal to RLMConfiguration.deleteRealmIfMigrationNeeded")
     }
+    
+    func testAppCredentialSupport() {
+
+        XCTAssertEqual(ObjectiveCSupport.convert(object: Credentials.facebook(accessToken: "accessToken")), RLMCredentials.init(facebookToken: "accessToken"))
```
Comment (tgoyne) [inline RealmSwift/Tests/ObjectiveCSupportTests.swift:101]: `.init` in all of these is redundant.

File: RealmSwift/Sync.swift [hunk 2]
```diff
@@ -575,9 +581,12 @@ public extension User {
     /// with the client from which it was created. On success a new user will be returned with the new linked credentials.
     /// @param credentials The `Credentials` used to link the user to a new identity.
     /// @returns A publisher that eventually return `Result.success` or `Error`.
-    func linkUser(with credentials: Credentials) -> Future<User, Error> {
+    func linkUser(credentials: Credentials) -> Future<User, Error> {
         return Future { promise in
-            self.linkUser(with: credentials) { user, error in
+            // ObjectiveCSupport.convert(object: credentials)
+//            self.linkUser(with: credentials) { user, error in
```
Comment (tgoyne) [inline RealmSwift/Sync.swift:587]: Leftover commented-out code

File: Realm/ObjectServerTests/SwiftObjectServerTests.swift [hunk 3]
```diff
@@ -741,8 +697,8 @@ class SwiftObjectServerTests: SwiftSyncTestCase {
 
         let loginEx = expectation(description: "Login user")
         var syncUser: User?
-
-        app.login(credentials: Credentials(email: email, password: password)) { (user, error) in
+
```
Comment (tgoyne) [inline Realm/ObjectServerTests/SwiftObjectServerTests.swift:696]: trailing whitespace

**Review by tgoyne — COMMENTED** [review]
enum is short for "enumeration" and not "enumerable", but also should just be written as "enum" even in prose.

File: RealmSwift/App.swift [hunk 4]
```diff
@@ -74,14 +74,46 @@ public typealias PushClient = RLMPushClient
...
+    /// Credentials from a Facebook access token.
+    case facebook(accessToken: String)
+    /// Credentials from a Google serverAuthCode.
+    case google(serverAuthCode: String)
+    /// Credentials from an Apple id token.
+    case apple(idToken: String)
+    /// Credentials from an email and password.
+    case emailPassword(email: String, password: String)
+    /// Credentials from a JSON Web Token
+    case JWT(token: String)
```
Comment (jsflax) [inline RealmSwift/App.swift:94]: This should be lowercase, thoughts @tgoyne ?

File: Realm/ObjectServerTests/RLMObjectServerTests.mm [hunk 5]
```diff
@@ -487,8 +487,7 @@ - (void)testAppleCredential {
 
 - (void)testFunctionCredential {
     NSError *error;
-    RLMCredentials *functionCredential = [RLMCredentials credentialsWithFunctionPayload:@{ @"dog" : @{ @"name" : @"fido" } }
-                                                                                        error:&error];
+    RLMCredentials *functionCredential = [RLMCredentials credentialsWithFunctionPayload:@{ @"dog" : @{ @"name" : @"fido" } }];
```
Comment (jsflax) [inline Realm/ObjectServerTests/RLMObjectServerTests.mm:490]: extra spaces in dictionary literal:
```objc
@{@"dog": @{@"name": @"fido"}}
```

File: RealmSwift/App.swift (hunk 4 above)
Comment (ericjordanmossman) [reply RealmSwift/App.swift:94]: as in `case jwt(token: String)`?

Comment (tgoyne) [reply RealmSwift/App.swift:94]: Yes, since it's the first word in the name it's all lowercase.

**Comment by ericjordanmossman** [comment]
I also need to merge in v10 changes before taking out of draft.

**Review by jsflax — APPROVED** [review]

**Review by tgoyne — APPROVED** [review]

**Pull request merged**